File: lib/models/telemetry.dart
// lib/models/telemetry.dart

class CpuData {
  final double percent;
  final List<double> perCore;
  final int coreCount;
  final double freqMhz, freqMaxMhz;
  final double load1, load5, load15;

  const CpuData({
    required this.percent, required this.perCore, required this.coreCount,
    required this.freqMhz, required this.freqMaxMhz,
    required this.load1, required this.load5, required this.load15,
  });

  factory CpuData.fromJson(Map<String, dynamic> j) => CpuData(
    percent:    (j['percent']      as num).toDouble(),
    perCore:    (j['per_core']  as List).map((e) => (e as num).toDouble()).toList(),
    coreCount:  (j['core_count']   as num).toInt(),
    freqMhz:    (j['freq_mhz']     as num).toDouble(),
    freqMaxMhz: (j['freq_max_mhz'] as num? ?? 0).toDouble(),
    load1:      (j['load_1']       as num).toDouble(),
    load5:      (j['load_5']       as num).toDouble(),
    load15:     (j['load_15']      as num).toDouble(),
  );
}

class RamData {
  final double totalMb, usedMb, availableMb, cachedMb, buffersMb, percent;

  const RamData({
    required this.totalMb, required this.usedMb, required this.availableMb,
    required this.cachedMb, required this.buffersMb, required this.percent,
  });

  factory RamData.fromJson(Map<String, dynamic> j) => RamData(
    totalMb:    (j['total_mb']     as num).toDouble(),
    usedMb:     (j['used_mb']      as num).toDouble(),
    availableMb:(j['available_mb'] as num).toDouble(),
    cachedMb:   (j['cached_mb']    as num? ?? 0).toDouble(),
    buffersMb:  (j['buffers_mb']   as num? ?? 0).toDouble(),
    percent:    (j['percent']      as num).toDouble(),
  );
}

class SwapData {
  final double totalMb, usedMb, percent;
  const SwapData({required this.totalMb, required this.usedMb, required this.percent});
  factory SwapData.fromJson(Map<String, dynamic> j) => SwapData(
    totalMb: (j['total_mb'] as num).toDouble(),
    usedMb:  (j['used_mb']  as num).toDouble(),
    percent: (j['percent']  as num).toDouble(),
  );
}

class DiskData {
  final String mount, device, fstype;
  final double totalGb, usedGb, freeGb, percent, readBps, writeBps;

  const DiskData({
    required this.mount, required this.device, required this.fstype,
    required this.totalGb, required this.usedGb, required this.freeGb,
    required this.percent, required this.readBps, required this.writeBps,
  });

  factory DiskData.fromJson(Map<String, dynamic> j) => DiskData(
    mount:    j['mount']    as String? ?? '',
    device:   j['device']   as String? ?? '',
    fstype:   j['fstype']   as String? ?? '',
    totalGb:  (j['total_gb'] as num).toDouble(),
    usedGb:   (j['used_gb']  as num).toDouble(),
    freeGb:   (j['free_gb']  as num).toDouble(),
    percent:  (j['percent']  as num).toDouble(),
    readBps:  (j['read_bps']  as num? ?? 0).toDouble(),
    writeBps: (j['write_bps'] as num? ?? 0).toDouble(),
  );
}

class NetData {
  final String iface;
  final double bytesSentS, bytesRecvS;
  final int packetsSent, packetsRecv, errin, errout;

  const NetData({
    required this.iface, required this.bytesSentS, required this.bytesRecvS,
    required this.packetsSent, required this.packetsRecv,
    required this.errin, required this.errout,
  });

  factory NetData.fromJson(Map<String, dynamic> j) => NetData(
    iface:       j['iface']         as String? ?? '',
    bytesSentS:  (j['bytes_sent_s'] as num).toDouble(),
    bytesRecvS:  (j['bytes_recv_s'] as num).toDouble(),
    packetsSent: (j['packets_sent'] as num? ?? 0).toInt(),
    packetsRecv: (j['packets_recv'] as num? ?? 0).toInt(),
    errin:       (j['errin']        as num? ?? 0).toInt(),
    errout:      (j['errout']       as num? ?? 0).toInt(),
  );
}

class FanData {
  final int rpm, thresholdC;
  final double apuTempC;
  const FanData({required this.rpm, required this.thresholdC, required this.apuTempC});
  factory FanData.fromJson(Map<String, dynamic> j) => FanData(
    rpm:        (j['rpm']         as num).toInt(),
    thresholdC: (j['threshold_c'] as num).toInt(),
    apuTempC:   (j['apu_temp_c']  as num).toDouble(),
  );
}

class TunnelStatus {
  final String state;
  final String? url;
  const TunnelStatus({required this.state, this.url});
  factory TunnelStatus.fromJson(Map<String, dynamic> j) => TunnelStatus(
    state: j['state'] as String? ?? 'stopped',
    url:   j['url']   as String?,
  );
  bool get isRunning => state == 'running';
}

class TelemetryFrame {
  final double ts;
  final FanData? fan;
  final CpuData? cpu;
  final RamData? ram;
  final SwapData? swap;
  final List<DiskData> disk;
  final List<NetData> net;
  final int uptimeS;
  final TunnelStatus? tunnel;
  final String? error;

  const TelemetryFrame({
    required this.ts, this.fan, this.cpu, this.ram, this.swap,
    this.disk = const [], this.net = const [],
    this.uptimeS = 0, this.tunnel, this.error,
  });

  bool get isError => error != null;

  factory TelemetryFrame.fromJson(Map<String, dynamic> j) {
    if (j.containsKey('error')) {
      return TelemetryFrame(ts: (j['ts'] as num).toDouble(), error: j['error'] as String);
    }
    return TelemetryFrame(
      ts:      (j['ts']       as num).toDouble(),
      fan:     j['fan']  != null ? FanData.fromJson(j['fan'])   : null,
      cpu:     j['cpu']  != null ? CpuData.fromJson(j['cpu'])   : null,
      ram:     j['ram']  != null ? RamData.fromJson(j['ram'])   : null,
      swap:    j['swap'] != null ? SwapData.fromJson(j['swap']) : null,
      disk:    (j['disk'] as List? ?? []).map((e) => DiskData.fromJson(e)).toList(),
      net:     (j['net']  as List? ?? []).map((e) => NetData.fromJson(e)).toList(),
      uptimeS: (j['uptime_s'] as num? ?? 0).toInt(),
      tunnel:  j['tunnel'] != null ? TunnelStatus.fromJson(j['tunnel']) : null,
    );
  }

  String get uptimeFormatted {
    final d = uptimeS ~/ 86400;
    final h = (uptimeS % 86400) ~/ 3600;
    final m = (uptimeS % 3600)  ~/ 60;
    if (d > 0) return '${d}d ${h}h ${m}m';
    if (h > 0) return '${h}h ${m}m';
    return '${m}m';
  }

  NetData? get primaryNet {
    if (net.isEmpty) return null;
    return net.firstWhere(
      (n) => n.iface != 'lo' && (n.bytesSentS > 0 || n.bytesRecvS > 0),
      orElse: () => net.firstWhere((n) => n.iface != 'lo', orElse: () => net.first),
    );
  }
}

File: StrawberryManager-iOS/Models/Telemetry.swift
// Telemetry.swift
// Telemetry data models matching backend API

import Foundation

struct TelemetryFrame: Codable, Equatable {
    let ts: Double
    let fan: FanData?
    let cpu: CPUData?
    let ram: RAMData?
    let swap: SwapData?
    let disk: [DiskData]
    let net: [NetData]
    let uptimeS: Int
    let tunnel: TunnelStatus?
    let error: String?
    
    enum CodingKeys: String, CodingKey {
        case ts, fan, cpu, ram, swap, disk, net
        case uptimeS = "uptime_s"
        case tunnel, error
    }
    
    var isError: Bool { error != nil }
    
    var uptimeFormatted: String {
        let days = uptimeS / 86400
        let hours = (uptimeS % 86400) / 3600
        let minutes = (uptimeS % 3600) / 60
        
        if days > 0 {
            return "\(days)d \(hours)h \(minutes)m"
        } else if hours > 0 {
            return "\(hours)h \(minutes)m"
        } else {
            return "\(minutes)m"
        }
    }
    
    var primaryNet: NetData? {
        guard !net.isEmpty else { return nil }
        // Find first non-loopback interface with traffic
        return net.first { $0.iface != "lo" && ($0.bytesSentS > 0 || $0.bytesRecvS > 0) }
            ?? net.first { $0.iface != "lo" }
            ?? net.first
    }
}

struct CPUData: Codable, Equatable {
    let percent: Double
    let perCore: [Double]
    let coreCount: Int
    let freqMhz: Double
    let freqMaxMhz: Double
    let load1: Double
    let load5: Double
    let load15: Double
    
    enum CodingKeys: String, CodingKey {
        case percent
        case perCore = "per_core"
        case coreCount = "core_count"
        case freqMhz = "freq_mhz"
        case freqMaxMhz = "freq_max_mhz"
        case load1 = "load_1"
        case load5 = "load_5"
        case load15 = "load_15"
    }
}

struct RAMData: Codable, Equatable {
    let totalMb: Double
    let usedMb: Double
    let availableMb: Double
    let cachedMb: Double
    let buffersMb: Double
    let percent: Double
    
    enum CodingKeys: String, CodingKey {
        case totalMb = "total_mb"
        case usedMb = "used_mb"
        case availableMb = "available_mb"
        case cachedMb = "cached_mb"
        case buffersMb = "buffers_mb"
        case percent
    }
}

struct SwapData: Codable, Equatable {
    let totalMb: Double
    let usedMb: Double
    let percent: Double
    
    enum CodingKeys: String, CodingKey {
        case totalMb = "total_mb"
        case usedMb = "used_mb"
        case percent
    }
}

struct DiskData: Codable, Equatable {
    let mount: String
    let device: String
    let fstype: String
    let totalGb: Double
    let usedGb: Double
    let freeGb: Double
    let percent: Double
    let readBps: Double
    let writeBps: Double
    
    enum CodingKeys: String, CodingKey {
        case mount, device, fstype
        case totalGb = "total_gb"
        case usedGb = "used_gb"
        case freeGb = "free_gb"
        case percent
        case readBps = "read_bps"
        case writeBps = "write_bps"
    }
}

struct NetData: Codable, Equatable {
    let iface: String
    let bytesSentS: Double
    let bytesRecvS: Double
    let packetsSent: Int
    let packetsRecv: Int
    let errin: Int
    let errout: Int
    
    enum CodingKeys: String, CodingKey {
        case iface
        case bytesSentS = "bytes_sent_s"
        case bytesRecvS = "bytes_recv_s"
        case packetsSent = "packets_sent"
        case packetsRecv = "packets_recv"
        case errin, errout
    }
}

struct FanData: Codable, Equatable {
    let rpm: Int
    let thresholdC: Int
    let apuTempC: Double
    
    enum CodingKeys: String, CodingKey {
        case rpm
        case thresholdC = "threshold_c"
        case apuTempC = "apu_temp_c"
    }
}

struct TunnelStatus: Codable, Equatable {
    let state: String
    let url: String?
    
    var isRunning: Bool { state == "running" }
}
